ably/options.go

```diff
@@ -14,36 +15,86 @@ const (
 	ProtocolMsgPack = "msgpack"
 )
 
+var errInvalidKey = errors.New("invalid key format")
+
+var DefaultOptions = &ClientOptions{
+	RestEndpoint:     "rest.ably.io",
+	RealtimeEndpoint: "realtime.ably.io",
+	Protocol:         ProtocolJSON, // TODO: make it ProtocolMsgPack
+}
+
 type ClientOptions struct {
-	RealtimeEndpoint string
 	RestEndpoint     string
-	Key              string
-	Token            string
-	Secret           string
+	RealtimeEndpoint string
+	KeyName          string
+	KeySecret        string
 	ClientID         string
-	Protocol         string
+	Protocol         string // either ProtocolJSON or ProtocolMsgPack
 	UseTokenAuth     bool
 	NoTLS            bool
 
 	HTTPClient *http.Client
 }
 
-func (p *ClientOptions) Prepare() {
-	if p.Key != "" {
-		p.parseKey()
+func NewClientOptions(key string) (*ClientOptions, error) {
+	opts := &ClientOptions{}
+	if err := opts.SetKey(key); err != nil {
+		return nil, err
+	}
+	return opts, nil
+}
+
+func (opts *ClientOptions) restURL() string {
+	host := opts.RestEndpoint
+	if host == "" {
+		host = DefaultOptions.RestEndpoint
+	}
+	if opts.NoTLS {
+		return "http://" + host
+	}
+	return "https://" + host
+}
+
+func (opts *ClientOptions) realtimeURL() string {
+	host := opts.RealtimeEndpoint
+	if host == "" {
+		host = DefaultOptions.RealtimeEndpoint
+	}
+	if opts.NoTLS {
+		return "ws://" + host + ":80"
+	}
+	return "wss://" + host + ":443"
+}
+
+func (opts *ClientOptions) SetKey(key string) error {
```

Comment:
But then *ClientOptions should not be exported / accessible from `RestClient`, and right now it is; I'd love to unexport it, since once ClientOptions were passed from outside to client, they should also be accessed from outside and not from the `RestClient` - thanks to this change user would have no opportunity of altering this field; however I've seen Java library makes this field public. So the question is - are those ClientOptions needed to be accessbile from RestClient? If not, the it solves all the problems.